rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -129,254 +113,136 @@ impl SerialSubmitter {
-                return Ok(false);
+        match op.submit().await {
+            TxRunResult::Success => {
+                self.metrics.txs_submitted.inc();
+                self.validation_queue.push(Reverse(op));
             }
-        };
-
-        // If the gas payment requirement hasn't been met, move to the next tick.
-        let Some(gas_limit) = self
-            .gas_payment_enforcer
-            .message_meets_gas_payment_requirement(&msg.message, &tx_cost_estimate)
-            .await?
-        else {
-            info!(?tx_cost_estimate, "Gas payment requirement not met yet");
-            return Ok(false);
-        };
-
-        // Go ahead and attempt processing of message to destination chain.
-        debug!(?gas_limit, "Ready to process message");
-
-        // TODO: consider differentiating types of processing errors, and pushing to the
-        //  front of the run queue for intermittent types of errors that can
-        //  occur even if a message's processing isn't reverting, e.g. timeouts
-        //  or txs being dropped from the mempool. To avoid consistently retrying
-        //  only these messages, the number of retries could be considered.
-
-        let gas_limit = tx_cost_estimate.gas_limit;
-
-        if let Some(max_limit) = self.transaction_gas_limit {
-            if gas_limit > max_limit {
-                info!("Message delivery estimated gas exceeds max gas limit");
-                return Ok(false);
+            TxRunResult::DoNotRetry => {
+                self.metrics.txs_submitted.inc();
+            }
+            TxRunResult::Retry => {
+                self.metrics.txs_failed.inc();
+                self.run_queue.push(Reverse(op));
             }
+            TxRunResult::CriticalFailure(e) => return Err(e),
         }
 
-        // We use the estimated gas limit from the prior call to
-        // `process_estimate_costs` to avoid a second gas estimation.
-        let outcome = self
-            .mailbox
-            .process(&msg.message, &metadata, Some(gas_limit))
-            .await?;
-
-        // TODO(trevor): Instead of immediately marking as processed, move to a
-        //  verification queue, which will wait for finality and indexing by the
-        //  mailbox indexer and then mark as processed (or eventually retry if
-        //  no confirmation is ever seen).
+        Ok(())
+    }
 
-        self.gas_payment_enforcer
-            .record_tx_outcome(&msg.message, outcome)?;
-        if outcome.executed {
-            info!(
-                hash=?outcome.txid,
-                rq_sz=?self.run_queue.len(),
-                "Message successfully processed by transaction"
-            );
-            Ok(true)
-        } else {
-            info!(
-                hash=?outcome.txid,
-                "Transaction attempting to process transaction reverted"
-            );
-            Ok(false)
+    /// Validate submitted operations.
+    async fn tick_validate(&mut self) -> Result<()> {
```

Comment:
I wonder if we can make things even more generic (and if that's a good idea).

For example, you could imagine three queues of PendingMessages:
1. prepare_queue
2. submit_queue
3. confirm_queue

Each queue implements the PendingMessageQueue trait, which has:
```
    async fn pop(&mut self) -> PendingMessage, PopResult;
```

where:
```
pub enum PopResult {
    /// PendingMessage should be moved to a different queue, e.g...
    /// Successful preparation: prepare -> submit
    /// Successful submission: submit -> confirm
    /// Failed submission: submit -> prepare
    /// Failed confirmation: confirm -> prepare
    Move(PendingMessageQueue),
    
    /// PendingMessage should be dropped entirely, e.g....
    /// prepare queue checks and it's already been delivered
    /// confirmation queue checks and it's been confirmed 
    Drop,
    /// A retry-able error occurred and we should push to the back of
    /// the queue to retry later, e.g....
    /// Prepare queue unable to assemble metadata
    /// Prepare queue failed gas payment enforcement
    Retry (or "Push"),
    /// Pass the error up the chain, this is non-recoverable and indicates a
    /// system failure.
    CriticalFailure(Report),
}